pkg/component/controller/workerconfig/reconciler.go

```diff
@@ -0,0 +1,539 @@
+/*
+Copyright 2022 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workerconfig
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/v1beta1"
+	"github.com/k0sproject/k0s/pkg/applier"
+	"github.com/k0sproject/k0s/pkg/component/controller/leaderelector"
+	"github.com/k0sproject/k0s/pkg/component/manager"
+	workerconfig "github.com/k0sproject/k0s/pkg/component/worker/config"
+	"github.com/k0sproject/k0s/pkg/constant"
+	kubeutil "github.com/k0sproject/k0s/pkg/kubernetes"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubeletv1beta1 "k8s.io/kubelet/config/v1beta1"
+	"k8s.io/utils/pointer"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+	"sigs.k8s.io/yaml"
+)
+
+type resources = []*unstructured.Unstructured
+
+// Reconciler maintains ConfigMaps that hold configuration to be
+// used on k0s worker nodes, depending on their selected worker profile.
+type Reconciler struct {
+	log logrus.FieldLogger
+
+	clusterDomain string
+	clusterDNSIP  net.IP
+	clientFactory kubeutil.ClientFactoryInterface
+	leaderElector leaderelector.Interface
+
+	mu    sync.Mutex
+	state any
```

Comment:
I still have concerns about untyped state.
Basically what we are trying to do here is somehow implement union type in the language which has no union types. Trying to implement feature which do not exists in the language is always fishy to me and IMO leads to over engineered code. 

My suggestion here just to use type state string and use it for explicit compare like 
```
const (
stateStopped = "stateStopped"
)


if r.state == stateStopped {
    go t.requestStop() // or somehow else
}
```


That would require somehow simplify the code at line 363
```
	case reconcilerStarted:
			r.state = reconcilerStopped{state.stopped}
			go state.requestStop()
			return state.stopped, nil
		case reconcilerStopped:
			return state.stopped, nil
```

but from my perspective it would a beneficial to us, because it is hard to read now (I still not sure I fully understand it tbh).


Also, not related to this particular line the comment left for, but may I suggest if we start targetting less generic, but more specific type of a code? it is internal thing, we read it more often than write it, and I would suggest to try aiming for the simpler code.

CC: @makhov @jnummelin wdyt?

Reply:
Done. Using newtyped strings to track the current state now.